userapi/internal/api.go

```diff
@@ -414,3 +417,42 @@ func (a *UserInternalAPI) PerformAccountDeactivation(ctx context.Context, req *a
 	res.AccountDeactivated = err == nil
 	return err
 }
+
+// PerformOpenIDTokenCreation creates a new token that a relying party uses to authenticate a user
+func (a *UserInternalAPI) PerformOpenIDTokenCreation(ctx context.Context, req *api.PerformOpenIDTokenCreationRequest, res *api.PerformOpenIDTokenCreationResponse) error {
+	token := util.RandomString(24)
+	tokenHash := getTokenHash(token)
+	nowMS := time.Now().UnixNano() / int64(time.Millisecond)
+	expiresMS := nowMS + (3600 * 1000) // 60 minutes
+
+	err := a.AccountDB.CreateOpenIDToken(ctx, fmt.Sprint(tokenHash.Sum32()), req.UserID, expiresMS)
+
+	res.Token = api.OpenIDToken{
+		Token:     token,
+		UserID:    req.UserID,
+		ExpiresTS: expiresMS,
+	}
+
+	return err
+}
+
+// QueryOpenIDToken validates that the OpenID token was issued for the user, the replying party uses this for validation
+func (a *UserInternalAPI) QueryOpenIDToken(ctx context.Context, req *api.QueryOpenIDTokenRequest, res *api.QueryOpenIDTokenResponse) error {
+	tokenHash := getTokenHash(req.Token)
+	openIDTokenAttrs, err := a.AccountDB.GetOpenIDTokenAttributes(ctx, fmt.Sprint(tokenHash.Sum32()))
+	if err != nil {
+		return err
+	}
+
+	res.Sub = openIDTokenAttrs.UserID
+	res.ExpiresTS = openIDTokenAttrs.ExpiresTS
+
+	return nil
+}
+
+// used for getting the format in which the token is stored to prevent plaintext storage of sensitive info
+func getTokenHash(token string) hash.Hash32 {
+	tokenHash := fnv.New32a()
```

Comment:
Whilst I appreciate the thought, we shouldn't need to do this. There's several reasons here.

Firstly, if the database is compromised then an attacker can just query information directly. Given this token is only used to get the user ID, there's nothing of interest being protected here data wise.

Secondly, the point of hashing passwords is to prevent the password getting leaked in the event of a database compromise. Leaked passwords are valuable because people tend to reuse passwords between services, so there is intrinsic value to them beyond just access to the compromised service.

Thirdly, if you _are_ hashing passwords, you definitely don't want to use something like FNV which is designed for speed. Hashing once (even with something better suited like SHA256) is trivial to break. For example, Google search `000c285457fc971f862a79b786476c78812c8897063c6fa9c045f579a3b2d63f` and tell me what my password is :) This is why libraries and algorithms exist for securely storing passwords, the most popular being bcrypt and scrypt. These algorithms hash many times, include salt (extra data to make sure 2 people with the same password don't get the same end value), etc. These algorithms are designed to be computationally (and in scrypt's case, memory) expensive to avoid brute force attacks.

So in conclusion:
 - The token is not guarding anything sensitive.
 - The token itself is not valuable.

So please just insert the token directly into the database.

Reply:
All good point, I just switched to plaintext storage.